dell-bios-fan-control: init package and module

rickyelopez opened the pull request:
Adds the `dell-bios-fan-control` package and matching one-shot systemd service which runs on multi-user.target.

##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] `meta.mainProgram` is set, if applicable.
- [ ] ~build time only dependencies are declared in `nativeBuildInputs`~
- [x] source is fetched from an official or trusted location
- [x] source is fetched using the appropriate function
- [ ] ~the motives for any special packaging choices are documented~
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [ ] ~patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed~
- [ ] ~patches that are remotely available are fetched rather than vendored~

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rickyelopez committed:
dell-bios-fan-control: init package

rickyelopez committed:
maintainers: add rickyelopez

ethancedwards8 commented on pkgs/by-name/de/dell-bios-fan-control/package.nix:
```diff
@@ -0,0 +1,35 @@
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp dell-bios-fan-control $out/bin/dell-bios-fan-control
```
could probably just use `installBin` here

rickyelopez replied:
nice, didn't know that existed, thanks!

acid-bong reviewed (changes requested):
A general must-have: a module and a package have to be in separate commits, first package, then module

rickyelopez committed:
dell-bios-fan-control: init module

acid-bong commented:
No need to delete the module, you can still add it here, just a separate third commit after the package one (separate commit, not a separate PR)

acid-bong reviewed (approved):
I'll only approve the package part, as I have no idea how to set up such tools in modules (nor am i super good in systemd unit shenanigans 🙈).

One last request tho: append ` and module` to the title

rickyelopez commented:
> I'll only approve the package part, as I have no idea how to set up such tools in modules

Sounds good, thanks! Should I find somebody else to review the module? If so, what's the best way for me to do that?

h7x4 commented on nixos/modules/services/hardware/dell-bios-fan-control.nix:
```diff
@@ -0,0 +1,47 @@
+    systemd.services.dell-bios-fan-control-resume = {
+      description = "Restart dell-bios-fan-control on resume.";
+      wants = [ "dell-bios-fan-control.service" ];
+      wantedBy = [ "suspend.target" ];
```
I don't think I understand what's going on here. The description claims that `dell-bios-fan-control.service` gets restarted when the system resumes operation. Is the idea that whenever the system suspends, it triggers a sleep and then suspends so that when it wakes up, the service will be restarted? Do you happen to know why the service even needs a restart?

No worries if you're not sure and just copying what you're seeing in AUR, I'm just curious :)

h7x4 replied:
I suppose this is the answer: https://aur.archlinux.org/packages/dell-bios-fan-control-git#comment-939247

It would be kinda nice to find out what causes the state to reset and make a more dependable target, but I guess that is out of scope of this PR. Maybe you could leave a comment pointing to this discussion though, so anyone reading the code wouldn't have to dig it up?

h7x4 commented on nixos/modules/services/hardware/dell-bios-fan-control.nix:
```diff
@@ -0,0 +1,48 @@
+      description = "Disables BIOS control of fans at boot.";
+      wants = [ "dell-bios-fan-control-resume.service" ];
+      wantedBy = [ "multi-user.target" ];
+      before = [ "i8kmon.service" ];
```
I don't think we have this service in nixpkgs? I suppose the kernel module is still needed for this to work though. There is a shiny new condition for kernel modules in systemd v258 that we could use.

```suggestion
      unitConfig.ConditionKernelModuleLoaded = "i8k";
```

h7x4 commented on nixos/modules/module-list.nix:
```diff
@@ -648,6 +648,7 @@
   ./services/hardware/buffyboard.nix
   ./services/hardware/ddccontrol.nix
   ./services/hardware/deepcool-digital-linux.nix
+  ./services/hardware/dell-bios-fan-control.nix
```
Could you add a release note?

rickyelopez replied:
sure I can do that. Forgive my ignorance, but which set of release notes does this belong in? Does this go straight into the newest (rl-2605.md)?

rickyelopez commented:
@h7x4 addressed all comments and added release notes, could you please take another look? Lmk if I need to add more/different details to the release notes! Thanks!

rickyelopez commented:
hey @h7x4 sorry for the double ping but you mentioned that today was the last day to merge this before the 25.11 cut. Can you take a look? No worries if not, I can move the release notes to the other doc.
